### ISSUE TITLE
BUG: The BatchNorm layer is not deserialized well.

### ISSUE THREAD

**Issue by zsogitbe** [mlpack/mlpack#1504]
I was saving and then loading a trained ann network while testing and I have noticed that the network is serialized well but when deserialized the BatchNorm layer is not loaded well. After loading, the beta and gamma vector sizes were still the default (10) and the values were not loaded.

**Comment by zsogitbe** [comment]
I think that you will most probably need to do 'if (Archive::is_loading::value)' and 'set_size' for 'beta' and 'gamma' in the BatchNorm layer serialization! They are not deserialized/loaded well because the size is still the default... (the serialization/save is OK!). I am not sure how to get the right size at that place in the code because I am still new to mlpack but I am sure that more experienced users will know it immediately. I hope this helps!

**Comment by zsogitbe** [comment]
This pull request does not work either. I would like to politely suggest that you always test the code before committing a pull request in order to prevent the accidental degradation of the library.
I have also noticed that you have removed the serialization of beta and gamma entirely. Are you sure that you know what you are doing? 
I do not have time to help you further with this issue because I am new to the library and I am just testing it.

**Comment by ShikharJ** [comment]
Hmm, weird. @rcurtin Can you think of a reason why the serialization might not be working?

**Comment by rcurtin** [comment]
> Are you sure that you know what you are doing?

I'm sure that @ShikharJ knows what he is doing---he has made substantial contributions to the library in the past year.

In this situation Shikhar is right: the `gamma` and `bias` matrices are simply aliases on top of the `weights` matrix.  The ANN code is complex so it's not necessarily easy to read, and the way that the network interacts with each layer is also complex.  For the sake of serialization, it turns out that the network itself holds all the weights in each layer as a contiguous matrix and that will be serialized, so we don't need to do that in the layer itself.

However it seems like maybe there are some other elements that aren't being serialized: `size`, `eps`, and `stats` jump out at me as possibilities.  Another possibility is that `stats` is being reset in `Reset()`, but for the batch normalization layer we should not be resetting that information when we serialize/deserialize the matrix.  Instead I think that should only happen at initialization time.

@zsogitbe: you're right that the PR that Shikhar proposed is not tested.  (If it was, you would not have encountered the bug at all.  But perfect testing coverage is not necessarily possible with such complex software...)  I think that a nice thing to do might be to work in your failing code as a test case, if you can supply simple code to build the network and then the code after serialization that fails.

**Comment by zsogitbe** [comment]
Thank you for your answer and clarification Ryan! I am just starting to test mlpack. The BatchNorm layer does not improve my test cases but makes them worse, therefore I will not use it further for the moment. I just wanted to let you know that there is something wrong, I think that you are right about the eps or stats reset. I advice you (the one who will try to solve the problem) in this case to save the xml model and look at the output, this is already a first easy test. I am sorry that I can not help you more in this at this time. I do not know the software yet well. If I will decide that I will use it (after some tests) then I will definitely help later!
I understand your intentions and I also find it very nice that you protect Shikhar! I however think that you should have more discipline while developing the software to always make at least one test before committing it. This will also reduce the final bugs in the software and it is easier to test a small part than the whole complex thing! But I of course understand that there can be bugs in any software, so consider my remarks just as help.

**Comment by zsogitbe** [comment]
You can use a very simple example you probably have already for the testing. Just build a simple nn, train it, save it, load it and try to predict. You could also make an automatic test case from this which could for example compare the prediction results before and after loading the model. One BatchNorm layer would be sufficient for the test:

FFN <NegativeLogLikelihood<>, RandomInitialization> model;
model.Add<Linear<> >(trainX.n_rows, H1);
model.Add<BatchNorm<> >();
model.Add<ReLULayer<> >();
model.Add<Linear<> >(H1, H2);
model.Add<ReLULayer<> >();
model.Add<Linear<> >(H2, nClasses);
model.Add<LogSoftMax<> >();

SGD <AdamUpdate> optimizer(...

train the model CYCLES times

save the model with: mlpack::data::Save(...

load the model with: mlpack::data::Load(...

predict

**Comment by rcurtin** [comment]
Ok, I think that we can mark this as resolved since #1508 is merged now.  Thanks for the bug report!  :+1:
